### ISSUE TITLE
[BUG] Since 2.14.1 docker compose run no longer executes RUN commands in DockerFile before the command is run

### ISSUE THREAD

**Issue by martinburchell** [docker/compose#10220]
### Description

We have a [Python installer script](https://github.com/ucam-department-of-psychiatry/crate/blob/master/installer/installer.py), which is a wrapper to Docker Compose using [python-on-whales](https://github.com/gabrieldemarmiesse/python-on-whales). 

The main server application container is built from a [Dockerfile](https://github.com/ucam-department-of-psychiatry/crate/blob/master/docker/dockerfiles/crate.Dockerfile). This (amongst other things) creates a Python virtual environment. Before the server can start up successfully, there needs to be a config file in place in one of the volumes. The installer script creates the config file with `docker compose run` on the main server container. The script is run in the Python virtual environment.

Before Docker Compose 2.14.1 `docker compose run` would first execute the `RUN` commands in the Dockerfile to (amongst other things) create the virtual environment.  

Since 2.14.1 the `RUN` commands are no longer executed first and the command fails with `No such file or directory` because the Python virtual environment doesn't exist.

Full log is at https://github.com/ucam-department-of-psychiatry/crate/actions/runs/3989901188/jobs/6842885560 . I've attached a download of the logs below in case this disappears.

Excerpts are at https://github.com/ucam-department-of-psychiatry/crate/issues/110

Is this a bug or is our installer relying on some undocumented behaviour of Docker Compose?

I could try to identify the breaking change with git-bisect but I would need help with building Docker Compose from source.

### Steps To Reproduce

Reproduced by running the installer from the CRATE project.  If necessary I could try to produce a smaller failing test case.

See https://github.com/ucam-department-of-psychiatry/crate/blob/master/.github/workflows/installer.yml and related files.

### Compose Version

```Text
Anything >= 2.14.1
```

### Docker Environment

```Text
Client:
Context:    default
Debug Mode: false
Plugins:
app: Docker App (Docker Inc., v0.9.1-beta3)
buildx: Docker Buildx (Docker Inc., v0.10.0-docker)
compose: Docker Compose (Docker Inc., v2.15.1)
scan: Docker Scan (Docker Inc., v0.23.0)
WARNING: Plugin "/home/martinb/.docker/cli-plugins/docker-compose-v2.2.2" is not valid: plugin candidate "compose-v2.2.2" did not match "^[a-z][a-z0-9]*$"
WARNING: Plugin "/home/martinb/.docker/cli-plugins/docker-compose-v2.15.1" is not valid: plugin candidate "compose-v2.15.1" did not match "^[a-z][a-z0-9]*$"

Server:
Containers: 28
Running: 3
Paused: 0
Stopped: 25
Images: 111
Server Version: 20.10.23
Storage Driver: overlay2
Backing Filesystem: extfs
Supports d_type: true
Native Overlay Diff: true
userxattr: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Cgroup Version: 1
Plugins:
Volume: local
Network: bridge host ipvlan macvlan null overlay
Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
Swarm: inactive
Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 5b842e528e99d4d4c1686467debf2bd4b88ecd86
runc version: v1.1.4-0-g5fd4c4d
init version: de40ad0
Security Options:
apparmor
seccomp
Profile: default
Kernel Version: 5.15.0-56-generic
Operating System: Ubuntu 20.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 31.12GiB
Name: lap-l380-mb
ID: AJJN:F5C2:ENDS:M5E3:JZT7:K3XI:S6G6:PC7U:PWTK:JTWV:FI4Z:UCYL
Docker Root Dir: /var/lib/docker
Debug Mode: false
Username: crateanon
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
127.0.0.0/8
Live Restore Enabled: false

```
```


### Anything else?

Output of `docker version`
```
Client: Docker Engine - Community
 Version:           20.10.23
 API version:       1.41
 Go version:        go1.18.10
 Git commit:        7155243
 Built:             Thu Jan 19 17:36:25 2023
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.23
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.10
  Git commit:       6051f14
  Built:            Thu Jan 19 17:34:14 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.15
  GitCommit:        5b842e528e99d4d4c1686467debf2bd4b88ecd86
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

Output of `docker context show`:

```

Usage:  docker context COMMAND

Manage contexts

Commands:
  create      Create a context
  export      Export a context to a tar or kubeconfig file
  import      Import a context from a tar or zip file
  inspect     Display detailed information on one or more contexts
  ls          List contexts
  rm          Remove one or more contexts
  update      Update a context
  use         Set the current docker context

Run 'docker context COMMAND --help' for more information on a command.

```
[Full log of failure](https://github.com/docker/compose/files/10546004/logs_1287.zip)

**Comment by laurazard** [comment]
Hi @martinburchell, thanks for the report. `docker compose run` does not run things in existing containers – instead, it creates a one-off container and executes the specified command. If you want to execute a command in a running container, `docker compose exec` would be your best bet. Asides from this though, `RUN` instructions in Dockerfiles are part of the build process, and should run regardless, not at the container create/start time but at the image building time.

I tried looking through the repo you mentioned, but it would make debugging your issue a lot easier if you could point us to the specific Dockerfiles/Compose files being used, which RUN commands are not being executed, and where the `docker compose` commands are being called – or, even better, a small reproducible example.

Regardless, I have some doubts there's anything wrong with Dockerfile `RUN` commands not being executed, and there might be something else that changed in the meantime that might be breaking your workflow.

**Comment by martinburchell** [comment]
> Regardless, I have some doubts there's anything wrong with Dockerfile `RUN` commands not being executed, and there might be something else that changed in the meantime that might be breaking your workflow.

Yes my first attempt at building a minimal example didn't fail. I'll get back to you when I have something failing.

**Comment by martinburchell** [comment]
@laurazard take a look at the minimal example at https://github.com/martinburchell/docker-compose-10220

There are two GitHub actions pre and post v2.14.1. One fails, the other succeeds https://github.com/martinburchell/docker-compose-10220/actions

One weird thing is that both workflows pass if I rename the image in docker-compose.yaml from 'crate' to 'test' or anything else it seems.

**Comment by laurazard** [comment]
Thanks @martinburchell! I think I see what's happening!

There's an issue here where, when you specify `image: crate` in your Compose file, Compose is, instead of building the image specified in the Dockerfile through:
```yaml
build:
  context: ./
```
it's pulling an image tagged `crate` from Hub (probably https://hub.docker.com/_/crate).

This behavior is explained in https://github.com/compose-spec/compose-spec/blob/master/build.md#consistency-with-image. When both attributes are present, Compose will first try to pull the image defined in `image: xxx`, then build it if not available. To get around this, you can rename the image in `image`, or simply not set it.

So what's happening isn't that the image isn't being built correctly, or not running the `RUN` commands, simply that the image being run when you do `compose run` isn't the image defined in your Dockerfile, but the image being pulled defined in `image`.

I'm closing this issue since things seem to be behaving as expected. Please comment further if you have further information/questions.
